packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -378,31 +378,73 @@ export default class CodeGen {
 
-        return memorizationId;
+        return memoizationId;
     }
 
     genBooleanAttributeExpr(bindExpr: t.Expression) {
         return t.conditionalExpression(bindExpr, t.literal(''), t.literal(null));
     }
 
     genEventListeners(listeners: EventListener[]) {
-        const listenerObj = Object.fromEntries(
-            listeners.map((listener) => [listener.name, listener])
-        );
-        const listenerObjectAST = objectToAST(listenerObj, (key) => {
-            const componentHandler = this.bindExpression(listenerObj[key].handler);
-            const handler = this.genBind(componentHandler);
+        let hasLocalListeners = false;
 
-            return memorizeHandler(this, componentHandler, handler);
-        });
+        const listenerObj: Record<string, { handler: t.Expression; isLocal: boolean }> = {};
+
+        for (const { name, handler } of listeners) {
+            const componentHandler = this.bindExpression(handler) as t.MemberExpression;
+            const id = getMemberExpressionRoot(componentHandler);
+            const isLocal = this.isLocalIdentifier(id);
+
+            if (isLocal) {
+                hasLocalListeners = true;
+            }
 
-        return t.property(t.identifier('on'), listenerObjectAST);
+            listenerObj[name] = { handler: this.genBind(componentHandler), isLocal };
+        }
+
+        const memoize = (expr: t.Expression) => {
+            const memoizedId = this.getMemoizationId();
+            return t.logicalExpression(
+                '||',
+                memoizedId,
+                t.assignmentExpression(
+                    '=',
+                    t.memberExpression(t.identifier(TEMPLATE_PARAMS.CONTEXT), memoizedId),
+                    expr
+                )
+            );
+        };
+
+        const listenerObjAST = objectToAST(
+            listenerObj,
+            hasLocalListeners
+                ? // If there are local listeners, we need to memoize individual handlers
+                  // Example input: <template for:each={list} for:item="task">
+                  //                  <button onclick={task.delete}>[X]</button>
+                  //                </template>
+                  // Output: { click: api_bind(task.delete) }
+                  (k) => {
+                      const { isLocal, handler } = listenerObj[k];
+                      return isLocal ? handler : memoize(handler);
+                  }
+                : // If there are no local listeners, we can memoize the entire object
+                  // Input: <template>
+                  //          <button onclick={create}>New</button>
+                  //        </template>
+                  // Output: _m1 || ($ctx._m1 = { click: api_bind($cmp.create) })
+                  (k) => listenerObj[k].handler
+        );
+
+        return t.property(
+            t.identifier('on'),
+            hasLocalListeners ? listenerObjAST : memoize(listenerObjAST)
+        );
```

Comment:
Rather than checking `hasLocalListeners` twice and using ternary conditionals each time, I think the logic would be more readable with a more simple:

```js
if (hasLocalListeners) {
  // ...
} else {
  // ...
}
```